spacy_experimental/coref/coref_component.py

```diff
@@ -0,0 +1,395 @@
+from typing import Iterable, Tuple, Optional, Dict, Callable, Any, List
+import warnings
+
+from thinc.types import Floats2d, Floats3d, Ints2d
+from thinc.api import Model, Config, Optimizer
+from thinc.api import set_dropout_rate, to_categorical
+from itertools import islice
+from statistics import mean
+import srsly
+
+from spacy.pipeline.trainable_pipe import TrainablePipe
+from spacy.language import Language
+from spacy.training import Example, validate_examples, validate_get_examples
+from spacy.errors import Errors
+from spacy.tokens import Doc
+from spacy.vocab import Vocab
+from spacy.util import from_disk, from_bytes
+
+from .coref_util import create_gold_scores, MentionClusters, create_head_span_idxs
+from .coref_util import get_clusters_from_doc, get_predicted_clusters
+from .coref_util import DEFAULT_CLUSTER_PREFIX
+
+from .coref_scorer import score_coref_clusters
+
+
+default_config = """
+[model]
+@architectures = "spacy-experimental.Coref.v1"
+distance_embedding_size = 20
+hidden_size = 1024
+depth = 1
+dropout = 0.3
+antecedent_limit = 50
+antecedent_batch_size = 512
+
+[model.tok2vec]
+@architectures = "spacy.Tok2Vec.v2"
+
+[model.tok2vec.embed]
+@architectures = "spacy.MultiHashEmbed.v1"
+width = 64
+rows = [2000, 2000, 1000, 1000, 1000, 1000]
+attrs = ["ORTH", "LOWER", "PREFIX", "SUFFIX", "SHAPE", "ID"]
+include_static_vectors = false
+
+[model.tok2vec.encode]
+@architectures = "spacy.MaxoutWindowEncoder.v2"
+width = ${model.tok2vec.embed.width}
+window_size = 1
+maxout_pieces = 3
+depth = 2
+"""
+DEFAULT_COREF_MODEL = Config().from_str(default_config)["model"]
+
+
+def coref_scorer(examples: Iterable[Example], **kwargs) -> Dict[str, Any]:
+    return score_coref_clusters(examples, **kwargs)
+
+
+def make_coref_scorer():
+    return coref_scorer
+
+
+@Language.factory(
+    "experimental_coref",
+    assigns=["doc.spans"],
+    requires=["doc.spans"],
+    default_config={
+        "model": DEFAULT_COREF_MODEL,
+        "span_cluster_prefix": DEFAULT_CLUSTER_PREFIX,
+        "scorer": {"@scorers": "spacy-experimental.coref_scorer.v1"},
+    },
+    default_score_weights={"coref_f": 1.0, "coref_p": None, "coref_r": None},
```

Comment:
Do you want the default score weights to be additionally keyed on the span cluster prefix?

Reply:
I don't think so - I think running multiple coref components in one pipeline is pretty unusual, so the hardcoded default here should be clear. 

I also think the behavior around parametrized score weights is confusing, in that if you forget to adjust your config you get 0 performance for components that are actually working just fine. In a situation where changing the score weights is important because you might have multiple components, or what the score refers to is unclear, that can be a reasonable tradeoff, but for coref I think the downsides outweigh the benefits.